src/fmu/pem/pem_functions/estimate_saturated_rock.py

```diff
@@ -30,68 +37,281 @@ def estimate_saturated_rock(
+    its own rock physics model (Friable, Patchy Cement, Regression, T-Matrix). The
+    workflow per zone is:
+      1. Create zone-masked inputs (mask outside zone cells only for filtering).
+      2. Call the appropriate RPM wrapper which internally flattens inputs using
+         filter_and_one_dim, runs the physics, and restores to 3D using
+         reverse_filter_and_restore.
+      3. Copy computed data values back into the global 3D result grids for cells in
+         this zone. The original simulation mask (inactive cells) is preserved; no
+         per‑zone mask overwrites occur.
+
+    Notes:
+    - Intermediate zone masks are not propagated to final outputs; only the original
+      reservoir inactive mask (fipnum_param.mask) remains.
+    - NaN handling (invalid physics results) is deferred to later processing, not
+      adjusted here.
+
+    Args:
+        rock_matrix: zone-aware rock matrix configuration
+        sim_init: simulation model initial properties (contains porosity, vsh, etc.)
+        eff_pres: effective / formation / overburden pressure objects per time step
+        matrix_props: effective mineral properties (already estimated upstream)
+        fluid_props: effective fluid properties per time step
+        model_directory: directory for model-specific parameter files (T-Matrix)
+        fipnum_param: FIPNUM grid partition descriptor
+
+    Returns:
+        List of SaturatedRockProperties (one per time step)
+
+    Raises:
+        ValueError: invalid zone coverage or unknown RPM type
+    """
+    # Validate the zone/region partitioning for saturated rocks
+    validate_fipnum_pvtnum_groups(
+        input_strings=[zone.fipnum for zone in rock_matrix.zone_regions],
+        fipnum_grid=fipnum_param,
+    )
+
+    # Validate zone coverage
+    fipnum_strings: list[str] = [zone.fipnum for zone in rock_matrix.zone_regions]
+    validate_zone_coverage(fipnum_strings, fipnum_param, zone_name="FIPNUM")
+
+    # Get FIPNUM grid data and mask
+    fipnum_data = fipnum_param.data
+    fipnum_mask = (
+        fipnum_param.mask
+        if hasattr(fipnum_param, "mask")
+        else np.zeros_like(fipnum_data, dtype=bool)
+    )
+
+    # Number of time steps
+    num_time_steps = len(eff_pres)
+
+    # Initialize result grids for each time step
+    # We'll accumulate results per zone and merge them
+    sat_rock_props_list: list[SaturatedRockProperties] = []
+
+    for time_idx in range(num_time_steps):
+        # Initialize grids for this time step
+        vp_grid = np.ma.masked_array(
+            np.full(sim_init.poro.shape, np.nan, dtype=float), mask=fipnum_mask
+        )
+        vs_grid = np.ma.masked_array(
+            np.full(sim_init.poro.shape, np.nan, dtype=float), mask=fipnum_mask
+        )
+        density_grid = np.ma.masked_array(
+            np.full(sim_init.poro.shape, np.nan, dtype=float), mask=fipnum_mask
+        )
```

Comment:
These mask operation happens many time in the code. Perhaps centralize it to a function under e.g. utils.py? The advantange is that you can get a single line call; in addition the centralized could do some more validation, e.g. asserting that the fipnum_mask is "ok" (e.g. that not all values are masked). You can also put the stuff in line 86-91 into this function